### PR TITLE
adds connect_ex method just like the original socket module

### PR DESCRIPTION
**Pull request by lucaszanella**
Using connect(), I implemented connect_ex() so it returns error codes. The only cased where it won't return error codes is when the error is a socks error, not a socket error

### CODE REVIEW

File: socks.py [hunk 1]
```diff
@@ -834,29 +834,49 @@ def connect(self, dest_pair):
...
+                        raise error
             except ProxyError:
                 # Protocol error while negotiating with proxy
                 self.close()
                 raise
+                
+    @set_self_blocking
+    def connect_ex(self, dest_pair):
+        """ https://docs.python.org/3/library/socket.html#socket.socket.connect_ex
+        Like connect(address), but return an error indicator instead of raising an exception for errors returned by the C-level connect() call (other problems, such as “host not found,” can still raise exceptions).
```
Comment (Anorov) [inline socks.py]: This line contains "fancy" quotation marks. Can you please convert them to regular quotation marks (`"`)?

File: socks.py [hunk 2]
```diff
@@ -834,29 +834,49 @@ def connect(self, dest_pair):
...
-            proxy_addr, proxy_port = proxy_addr
-            proxy_server = "{0}:{1}".format(proxy_addr, proxy_port)
-            printable_type = PRINTABLE_PROXY_TYPES[proxy_type]
-
-            msg = "Error connecting to {0} proxy {1}".format(printable_type,
-                                                             proxy_server)
-            log.debug("%s due to: %s", msg, error)
-            raise ProxyConnectionError(msg, error)
+            if not catch_errors:
+                    proxy_addr, proxy_port = proxy_addr
```
Comment (Anorov) [inline socks.py]: This block and other blocks are indented with 8 spaces, but the existing code is indented with 4 spaces.

File: socks.py [hunk 3]
```diff
@@ -834,29 +834,49 @@ def connect(self, dest_pair):
...
+    @set_self_blocking
+    def connect_ex(self, dest_pair):
+        """ https://docs.python.org/3/library/socket.html#socket.socket.connect_ex
+        Like connect(address), but return an error indicator instead of raising an exception for errors returned by the C-level connect() call (other problems, such as “host not found,” can still raise exceptions).
+        """
+        try:
+                self.connect(dest_pair, catch_errors=True)
+                return 0
+        except OSError as e:
+                if e.errno: #If the error is numeric (socket errors are numeric), then return number as connect_ex expects. Otherwise raise the error again (socket timeout for example)
```
Comment (Anorov) [inline socks.py]: Could you move this comment to the line before, and break it up into multiple lines?

**Comment by lucaszanella** [comment]
I made the corrections. Please take a look now :)

**Pull request merged**